internal/raft/snap/snapshotter.go

```diff
@@ -0,0 +1,225 @@
+package snap
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/pkg/pbutil"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+const (
+	snapSuffix = ".snap"
+)
+
+var (
+	ErrNoSnapshot    = errors.New("snap: no available snapshot")
+	ErrEmptySnapshot = errors.New("snap: empty snapshot")
+	ErrInvalidSnapshot = errors.New("snap: invalid snapshot")
+
+	// A map of valid files that can be present in the snap folder.
+	validFiles = map[string]bool{
+		"db": true,
+	}
+)
+
+type Snapshotter struct {
+	dir string
+}
+
+func New(dir string) *Snapshotter {
+	return &Snapshotter{
+		dir: dir,
+	}
+}
+
+func (s *Snapshotter) SaveSnapStream(snapshot raftpb.Snapshot, stream io.Reader) error {
+	if raft.IsEmptySnap(snapshot) {
+		return nil
+	}
+	return s.saveStream(&snapshot, stream)
+}
+
+func (s *Snapshotter) SaveSnap(snapshot raftpb.Snapshot) error {
+	if raft.IsEmptySnap(snapshot) {
+		return nil
+	}
+	return s.save(&snapshot)
+}
+
+func (s *Snapshotter) saveStream(snapshot *raftpb.Snapshot, stream io.Reader) error {
+	fname := fmt.Sprintf("%016x-%016x%s", snapshot.Metadata.Term, snapshot.Metadata.Index, snapSuffix)
+	snapFile := filepath.Join(s.dir, fname)
+	err := writeAndSyncFile(snapFile, snapshot, stream, 0666)
+	if err != nil {
+		err1 := os.Remove(snapFile)
+		if err1 != nil {
+			log.Printf("ERROR - failed to remove broken snapshot file %s", snapFile)
+		}
+	}
+	return err
+}
+
+func (s *Snapshotter) save(snapshot *raftpb.Snapshot) error {
+	return s.saveStream(snapshot, bytes.NewReader(snapshot.Data))
+}
+
+func writeAndSyncFile(filename string, snapshot *raftpb.Snapshot, data io.Reader, perm os.FileMode) error {
+	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return err
+	}
+	snapBts := pbutil.MustMarshal(snapshot)
+	snapLen := make([]byte, 4)
+	binary.LittleEndian.PutUint32(snapLen, uint32(len(snapBts)))
+
+	snapBuff := bytes.Buffer{}
+	snapBuff.Write(snapLen)
+	snapBuff.Write(snapBts)
+
+	_, err = io.Copy(f, &snapBuff)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(f, data)
+	if err != nil {
+		return err
+	}
+
+	err = fileutil.Fsync(f)
+	if err1 := f.Close(); err == nil {
+		err = err1
+	}
+	return err
+}
+
+func (s *Snapshotter) Load() (*raftpb.Snapshot, io.ReadCloser, error) {
+	names, err := s.snapNames()
+	if err != nil {
+		return nil, nil, err
+	}
+	var (
+		snap *raftpb.Snapshot
+		data io.ReadCloser
+	)
+	for _, name := range names {
+		if snap, data, err = loadSnap(s.dir, name); err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return nil, nil, ErrNoSnapshot
+	}
+	return snap, data, nil
+}
```

Comment:
How would this work when the raft picks up the snapshot to be transferred to a new node?

Reply:
This code handles saving application/state machine snapshots alone and has no bearing on RAFT interactions. During bootstrap/re-join of a follower, the leader currently sends only the snapshotted WAL (since the last compaction) along with the entries until commit index. These commits are subsequently read and applied by the application/state machine. Please look `replicator.go#211` for this loop where Nexus sends these commits down to the state machine impl. Periodically, the application/state machine is also given a chance to snapshot its local state. This is usually done to checkpoint the application state and perhaps take backups, etc.

Reply:
Okay, One question, I still have is that, how does a new node catchup currently with other nodes? Because I have tested this flow and it does transfer snapshot to a new node. (even when we limit the num of wal logs)

Reply:
Yes, as mentioned before. This snapshot you refer to being transferred, this is not application data but the WAL itself. The receiving node replays all the commits in that WAL onto its local SM. In fact, if you refer to `replicator_test.go`, we actually have a test for this scenario of a fresh node joining the cluster.

Reply:
Yup figured that out. Added https://github.com/flipkart-incubator/nexus/pull/21 which verifies this fact. We would need to fix this issue though :)